text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,335 @@
+
+This approach also offers an approach that can be developed and applied incrementally, allowing for time to find coner cases not considered by this RFC
+
+> What other designs have been considered and what is the rationale for not choosing them?
+
+To the author of this RFC's knowledge, there are no other open designs, other than the use tools that wrap Cargo entirely, such as [xargo].
+
+[xargo]: https://github.com/japaric/xargo
+
+> What is the impact of not doing this?
+
+By not doing this, Rust will continue to be difficult to use for users and platforms "on the edge", such as new platform developers or embedded and WASM users.
+
+# Prior art
+[prior-art]: #prior-art
+
+* [RFC1133] - This RFC from 2015 proposed making cargo aware of std. I still need to review in more detail to find the parts and syntax that may solve some open questions.
+* [xargo] - This external tool was used to achieve a similar workflow as described above, limited to use with a nightly compiler
+* [Cargo Issue 5002] - This issue proposed a syntax for explicit dependency on std
+* [Cargo Issue 5003] - This issue discussed how to be backwards compatible with crates that don't explicitly mention std
+
+[RFC1133]: https://github.com/rust-lang/rfcs/pull/1133
+[Cargo Issue 5002]: https://github.com/rust-lang/cargo/issues/5002
+[Cargo Issue 5003]: https://github.com/rust-lang/cargo/issues/5003
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## How are dependencies (or non-dependency) on `core` and `std` specified?
+
+For example in a `no_core` or `no_std` crate, how would we tell Cargo **not** to build the `core` and/or `std` dependencies?
+
+## Should `std` be rebuilt if `core` is rebuilt?
+
+Is it necessary to rebuild `std` using the customized `core`, even if no changes to `std` are necessary?
+
+## Should Cargo obtain or verify the source code for `libcore` or `libstd`?
+
+Right now we depend on `rustup` to obtain the correct source code for these libraries, and we rely on the user not to tamper with the contents. Are these reasonable decisions?
+
+## Should the custom built `libcore` and `libstd` reside locally or globally?
+
+e.g., should the build artifacts be placed in `target/`, only usable by this project, or in `.cargo/`, to be possibly reused by multiple projects, if they happen to have the same settings?
+
+## How do we handle `libcore` and `libstd`'s `Cargo.lock` file?
+
+Right now these are built using the global lock file in `rust-lang/rust`. Should this always be true? How should Cargo handle this gracefully?
+
+## Should profile changes always prompt a rebuild of `core`/`std`?
+
+For example, if a user sets their debug build to use `opt-level = 'z'`, should this rebuild `core`/`std` to use that opt level? Or should an additional flag, such as `apply-to-sysroot` be required to opt-in to this behavior, unless otherwise needed?
```

Comment:
I'm a little confused on this point. Would it rebuild core/std even if they are not listed in `[dependencies]` or `[patch]`? I fear that always rebuilding when using non-default profile settings would be too much disruption. It would also be a little confusing, because there are multiple profiles, none of which match the settings used in the precompiled libraries. Perhaps core/std should only be rebuilt when they are explicitly listed in `Cargo.toml`?

Another potential issue is that the features to enable for libstd currently are driven by bootstrap's config.toml, so it may not be obvious to the user that they need to enable things like "backtrace" to have feature parity with the defaults (which change per platform).  How do you switch to a custom-built std that retains feature parity with the default distribution per platform?

Reply:
That is exactly the open question here. On one hand: if the user states they want "opt-level = 's'", and we can now give them that for libstd too, it would make sense to keep the total program size down.

On the other hand, this could surprise some users, as it could drastically increase a clean build time by rebuilding core and standard for them.

Good call regarding `config.toml`, I was not aware of how this mechanism worked. In `xargo`, you can set some (or all?) of these flags using a `feature = [ ... ]` syntax. I would guess we would need to expose these features as "Cargo Features" rather than config.toml features so the users may configure them, with `default-features` matching what the CI builds of `libstd` and `libcore` currently are.

Reply:
I think they are "Cargo Features", `config.toml` just drives the defaults.  This is done [here](https://github.com/rust-lang/rust/blob/0f88167f89fffe321590c5148f21b7d51d44388d/src/bootstrap/lib.rs#L501-L514).  

One thing I'm aware of that is not driven via features, and that is sanitizer support. That is done [here](https://github.com/rust-lang/rust/blob/0f88167f89fffe321590c5148f21b7d51d44388d/src/bootstrap/compile.rs#L167-L180). It looks like sanitizers are used on the linux builds, but I don't really know anything about them. Looks like it requires llvm?

Maybe the defaults could be captured somewhere when the distribution is made, and Cargo could read those (maybe in the "src" component?  maybe rewrite the `Cargo.toml`?)?

Reply:
It's a general question whether any dependency needs to be rebuilt for these non-abi-affecting switches, so as usual I hope for a Cargo solution that doesn't special-case `core` and `std`.